File: outputs/audits/auto_reference_20260711/qa-system/prompt_builder.py
"""
Prompt 组装器 - 将规则、GoodCase、BadCase 组装进 Prompt
"""
from typing import List, Dict


def format_rules_compact(rules: List[Dict]) -> str:
    """将结构化规则格式化为精简文本（给模型看的）"""
    lines = []
    for r in rules:
        severity_tag = {"严重": "🔴", "中等": "🟡", "轻微": "🟢"}.get(r["severity"], "")
        lines.append(f"{r['id']}. [{r['severity']}] {r['title']}")
        # 核心判断标准
        detail = r["detail"]
        # 提取关键句
        if len(detail) > 300:
            detail = detail[:300] + "..."
        lines.append(f"   {detail}")
    return "\n".join(lines)


def format_good_case(case: Dict, index: int) -> str:
    """格式化一条 GoodCase"""
    return f"""### 通过案例 {index + 1}
- **任务类型**: {case.get('task_type', '')}
- **领域**: {case.get('category', '')}
- **题目**: {case['title'][:800]}
- **判定**: ✅ 通过 —— 此题目真实、信息充分、可执行，无违规项
"""


def format_bad_case(case: Dict, index: int) -> str:
    """格式化一条 BadCase"""
    issues = case.get("issues", "")
    return f"""### 不通过案例 {index + 1}
- **任务类型**: {case.get('task_type', '')}
- **领域**: {case.get('category', '')}
- **题目**: {case['title'][:800]}
- **判定**: ❌ 不通过 —— 问题如下：
  {issues[:400]}
"""


def build_full_prompt(
    base_prompt: str,
    title: str,
    task_type: str,
    attachment_details: str = "",
    output: str = "",
    good_cases: List[Dict] = None,
    bad_cases: List[Dict] = None,
) -> str:
    """
    组装完整的质检 Prompt

    结构:
    1. System: 基础质检 Prompt（你的原文）
    2. Few-Shot: GoodCase + BadCase
    3. User: 待检数据
    """
    good_cases = good_cases or []
    bad_cases = bad_cases or []

    # ===== 构建 Few-Shot 部分 =====
    few_shot_parts = []

    if good_cases:
        few_shot_parts.append("## 📗 参考——正确通过的题目（你的判断应与此一致）\n")
        for i, case in enumerate(good_cases):
            few_shot_parts.append(format_good_case(case, i))

    if bad_cases:
        few_shot_parts.append("\n## 📕 参考——被判定不通过的题目（注意同类问题）\n")
        for i, case in enumerate(bad_cases):
            few_shot_parts.append(format_bad_case(case, i))

    few_shot_text = "\n".join(few_shot_parts) if few_shot_parts else ""

    # ===== 构建用户输入（待检数据）=====
    user_input = f"""【待检数据】
任务类型：{task_type}

【规则2/3唯一可检查文本：题目字段】
<TITLE_FIELD>
{title}
</TITLE_FIELD>

【参考信息：只能用于理解任务和判断附件/数据是否支撑任务】
附件原文摘录：{attachment_details or "（空）"}
产物内容列：{output or "（空）"}
"""

    # ===== 组装最终 Prompt =====
    # System = 基础规则 + Few-Shot
    system_content = base_prompt

    if few_shot_text:
        system_content += f"\n\n{few_shot_text}"

    return {
        "system": system_content,
        "user": user_input,
    }


def estimate_tokens(text: str) -> int:
    """估算 token 数（中文字符约 1.5 token/字，英文约 0.3 token/字）"""
    chinese_chars = len([c for c in text if '一' <= c <= '鿿'])
    other_chars = len(text) - chinese_chars
    return int(chinese_chars * 1.5 + other_chars * 0.3)

File: outputs/audits/auto_reference_20260711/qa-system/prompt_builder.js
/**
 * Prompt 组装器 - 同时使用 GoodCase 和 BadCase 做 Few-Shot
 *
 * 核心设计：
 * - GoodCase 作为"合格范例"，展示通过质检的题目长什么样
 * - BadCase 作为"踩坑警告"，展示真实违规模式
 * - 双向对比让模型更准确理解合格/不合格的边界
 * - 加反偏见指令：防止模型盲目模仿 GoodCase 或只因为比 BadCase 好就放过
 */
function formatGoodCase(caseItem, index) {
  return `### ✅ 合格范例 ${index + 1}
- **题目摘要**: ${(caseItem.title || '').substring(0, 400)}
- **任务类型**: ${caseItem.taskType || ''}
- **通过原因**: 该题目信息支撑充分、场景真实自然、产物要求适度，各方面均符合质检规范。
`;
}

function formatBadCase(caseItem, index) {
  const issues = (caseItem.issues || '').substring(0, 300);
  return `### ❌ 踩坑案例 ${index + 1}
- **题目摘要**: ${(caseItem.title || '').substring(0, 400)}
- **被打回的原因**: ${issues}
- **教训**: 以上问题模式如果在待检题目中出现，同样需要判不通过。
`;
}

function buildFullPrompt({
  basePrompt,
  title,
  taskType,
  attachmentDetails = '',
  output = '',
  goodCases = [],
  badCases = [],
}) {
  // ===== 同时使用 GoodCase 和 BadCase =====
  const fewShotParts = [];

  // -- GoodCase：展示什么是对的 --
  if (goodCases.length > 0) {
    fewShotParts.push(`
## ✅ 参考：以下是真实通过质检的合格题目
**这些案例展示了符合规范的题目长什么样。它们信息支撑充分、场景真实、产物要求合理，因此通过了质检。**
**注意：待检题目的内容/领域与这些范例不同是正常的，不能因此扣分。你只需要参考它们的"规范性"水平。**

`);
    goodCases.forEach((c, i) => {
      fewShotParts.push(formatGoodCase(c, i));
    });
  }

  // -- BadCase：展示什么是错的 --
  if (badCases.length > 0) {
    fewShotParts.push(`
## ❌ 重要：以下是真实被质检打回的题目案例
**这些案例展示了必须判不通过的典型违规模式。如果待检题目出现同类问题，必须报错，不要手软。**

`);
    badCases.forEach((c, i) => {
      fewShotParts.push(formatBadCase(c, i));
    });
  }

  // 双向反偏见指令
  if (goodCases.length > 0 || badCases.length > 0) {
    fewShotParts.push(`
---
**关键提醒（非常重要）**：
1. ${goodCases.length > 0 ? '✅ GoodCase 展示的是"合格标准"——待检题目需要在规范性上达到同等水平才算通过。' : ''}
2. ${badCases.length > 0 ? '❌ BadCase 展示的是"红线"——待检题目中出现同类违规必须判不通过。' : ''}
3. 不要因为待检题目「看起来比 BadCase 好一点」就放过 —— 只要有同类违规就必须报
4. 也不要因为待检题目「写得长/看起来专业」就默认通过 —— 长不等于合格
5. ${goodCases.length > 0 ? 'GoodCase 的领域/内容与待检题目不同是正常的——不要因为内容不同而扣分，只比较"规范性"。' : ''}
6. 记住：规则是硬约束，不是软建议。命中"必须判错"的条目一定要报。
`);
  }

  const fewShotText = fewShotParts.join('');

  // ===== 构建用户输入（待检数据）=====
  const userInput = `【待检数据】
任务类型：${taskType}

【规则2/3唯一可检查文本：题目字段】
<TITLE_FIELD>
${title}
</TITLE_FIELD>

【参考信息：只能用于理解任务和判断附件/数据是否支撑任务】
附件原文摘录：${attachmentDetails || '（空）'}
产物内容列：${output || '（空）'}
`;

  return {
    system: basePrompt + fewShotText,
    user: userInput,
  };
}

module.exports = { buildFullPrompt };
